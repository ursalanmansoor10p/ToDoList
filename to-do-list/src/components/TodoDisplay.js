import React from "react";

import Checkbox from "react-custom-checkbox";
import { FaTrashAlt } from "react-icons/fa";
export default function Todo(
  //Props that are handed down
  //todo is the collection database object
  //togglecomplete is the reference to the parent function
  //handleDelete is the reference to the parent function
  { todo, toggleComplete, handleDelete }
) {
  //have a state of all todo titles for conditional rendering
  const [newTitle, setNewTitle] = React.useState(todo.title);

  console.log("todo", todo);

  const handleChange = (e) => {
    e.preventDefault();
    if (todo.complete === true) {
      setNewTitle(todo.title);
    } else {
      todo.title = "";
      setNewTitle(e.target.value);
    }
  };

  return (
    // the display div is split into three things, check mark. display and
    //delete button
    <div className="to" data-testid="todo-display">
      <Checkbox
        data-testid="checkbox"
        checked={todo.completed}
        icon={<img src={require("./check.png")} style={{ width: 24 }} alt="" />}
        borderColor="rgba(0, 0, 0, 0.452)"
        size={18}
        onChange={() => toggleComplete(todo)}
      />
      {/* conditional renders if completed should be crossed out  */}
      <div data-testid="todo-title">
        {todo.completed === true ? <del>{newTitle}</del> : newTitle}
      </div>
      <div>
        <div>
          {/*  delete references the handle delete function from parent */}
          <button
            className="button-delete"
            onClick={() => handleDelete(todo.id)}
          >
            <FaTrashAlt />
          </button>
        </div>
      </div>
    </div>
  );
}
