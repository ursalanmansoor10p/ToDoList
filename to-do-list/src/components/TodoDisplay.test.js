import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import TodoDisplay from "./TodoDisplay";

const todo = { id: 1, title: "abc", completed: false };

describe("TodoDisplay Component", () => {
  test("should match snapshot", () => {
    const { container } = render(
      <TodoDisplay
        todo={todo}
        handleDelete={() => undefined}
        toggleComplete={() => undefined}
      />
    );
    expect(container).toMatchSnapshot();
  });

  test("should call toggleComplete function when checkbox is toggled", () => {
    const toggleComplete = jest.fn();
    render(
      <TodoDisplay
        todo={todo}
        handleDelete={() => undefined}
        toggleComplete={toggleComplete}
      />
    );
    const checkbox = screen.getByTestId("checkbox");
    expect(checkbox.checked).toBe(todo.completed);
    fireEvent.click(checkbox);
    expect(toggleComplete).toBeCalledWith(todo);
  });

  test("should render the todo title with del if the todo is completed", () => {
    render(
      <TodoDisplay
        todo={{ ...todo, completed: true }}
        handleDelete={() => undefined}
        toggleComplete={() => undefined}
      />
    );

    const div = screen.getByTestId("todo-title");
    expect(div.childElementCount).toBe(1);
    expect(div.innerHTML).toContain(todo.title);
  });

  test("should render the todo title directly if the todo is not completed", () => {
    render(
      <TodoDisplay
        todo={todo}
        handleDelete={() => undefined}
        toggleComplete={() => undefined}
      />
    );

    const div = screen.getByTestId("todo-title");
    expect(div.childElementCount).toBe(0);
    expect(div.innerHTML).toBe(todo.title);
  });

  test("should call handleDelete function when delete button is clicked", () => {
    const handleDelete = jest.fn();
    render(
      <TodoDisplay
        todo={todo}
        handleDelete={handleDelete}
        toggleComplete={() => undefined}
      />
    );
    const button = screen.getByRole("button");
    fireEvent.click(button);
    expect(handleDelete).toBeCalledWith(todo.id);
  });
});
