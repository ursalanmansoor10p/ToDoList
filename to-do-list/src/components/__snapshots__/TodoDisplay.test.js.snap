// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoDisplay Component should match snapshot 1`] = `
<div>
  <div
    class="to"
    data-testid="todo-display"
  >
    <label
      class=""
      style="display: flex; align-items: center;"
    >
      <span>
        <div
          class=""
          style="height: 18px; width: 18px; border-width: 2px; border-color: rgba(0, 0, 0, 0.452); border-style: solid; border-radius: 5px; display: flex; align-items: center; justify-content: center;"
        >
          <input
            data-testid="checkbox"
            hidden=""
            name=""
            type="checkbox"
            value=""
          />
        </div>
      </span>
    </label>
    <div
      data-testid="todo-title"
    >
      abc
    </div>
    <div>
      <div>
        <button
          class="button-delete"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M32 464a48 48 0 0 0 48 48h288a48 48 0 0 0 48-48V128H32zm272-256a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zm-96 0a16 16 0 0 1 32 0v224a16 16 0 0 1-32 0zM432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</div>
`;
